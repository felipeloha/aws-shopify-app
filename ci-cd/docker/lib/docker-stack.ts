import * as cdk from '@aws-cdk/core';
import codebuild = require('@aws-cdk/aws-codebuild');
import ecr = require('@aws-cdk/aws-ecr');
import { Role, ServicePrincipal, PolicyStatement } from '@aws-cdk/aws-iam';

export class DockerStack extends cdk.Stack {
  public static repoName = 'aws-shopify-app';
  public static projectName = `${DockerStack.repoName}-docker`;

  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const repository = new ecr.Repository(
      this, 
      DockerStack.repoName,
      {
        repositoryName: DockerStack.repoName,
      }
      );
    repository.addLifecycleRule(
      { 
        tagStatus: ecr.TagStatus.UNTAGGED, 
        maxImageAge: cdk.Duration.days(1) 
      }
    );
    //remove all PUL images after 30 days
    //TODO is 30 days too much?
    repository.addLifecycleRule(
      { 
        tagPrefixList: ['PUL', 'develop', 'master'], 
        maxImageAge: cdk.Duration.days(30) 
      }
    );
    
    cdk.Tags.of(repository).add('Name', DockerStack.repoName);

    const bbSource = codebuild.Source.bitBucket({
      owner: 'mybitbucket',
      repo: DockerStack.repoName,
      reportBuildStatus: false,
    });

    const codebuildRole = new Role(this, 'codebuildRole', {
      assumedBy: new ServicePrincipal('codebuild.amazonaws.com')
    });
    codebuildRole.addToPolicy(
      new PolicyStatement({
        resources: ['*'],
        actions: ['logs:CreateLogGroup', 'logs:CreateLogStream', 'logs:PutLogEvents'],
      })
    );
    codebuildRole.addToPolicy(
      new PolicyStatement({
        resources: ['*'],
        actions: [
          'ecr:GetAuthorizationToken',
          'ecr:BatchCheckLayerAvailability',
          'ecr:GetDownloadUrlForLayer',
          'ecr:GetRepositoryPolicy',
          'ecr:DescribeRepositories',
          'ecr:ListImages',
          'ecr:DescribeImages',
          'ecr:BatchGetImage',
          'ecr:PutImage',
          'ecr:InitiateLayerUpload',
          'ecr:UploadLayerPart',
          'ecr:CompleteLayerUpload'
        ],
      })
    );


    const codebuildProject = new codebuild.Project(this, 'Codebuild', {
      buildSpec: codebuild.BuildSpec.fromSourceFilename('ci-cd/docker/buildspec.yaml'),
      projectName: DockerStack.projectName,
      role: codebuildRole,
      source: bbSource,
      timeout: cdk.Duration.minutes(10),
      environment: {
        computeType: codebuild.ComputeType.SMALL,
        buildImage: codebuild.LinuxBuildImage.AMAZON_LINUX_2_3,
        privileged: true,
      },
    });
    cdk.Tags.of(codebuildProject).add('Name', `codebuild-${DockerStack.projectName}`);
  }
}
