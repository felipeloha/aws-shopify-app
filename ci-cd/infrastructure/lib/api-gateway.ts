import * as cdk from '@aws-cdk/core';
import * as acm from '@aws-cdk/aws-certificatemanager';
import * as ecs_patterns from '@aws-cdk/aws-ecs-patterns';
import * as route53 from '@aws-cdk/aws-route53';
import * as apigateway from '@aws-cdk/aws-apigateway';
import * as targets from '@aws-cdk/aws-route53-targets';
import * as cwLogs from '@aws-cdk/aws-logs';

const domain = "domain.com";

export default function createApiGateway(
  stack: any,
  stackName: string,
  subdomain: string,
  fargateService: ecs_patterns.ApplicationLoadBalancedFargateService,
) {
  const targetURL = "http://" + fargateService.listener.loadBalancer.loadBalancerDnsName;
  const hostedZone = route53.HostedZone.fromLookup(stack, 'Zone', { domainName: domain });

  const certificate = new acm.Certificate(stack, `${subdomain}-certificate`, {
    domainName: subdomain,
    validation: acm.CertificateValidation.fromDns(hostedZone),
  });

  cdk.Tags.of(certificate).add('Name', `${subdomain}-certificate`);

  const apiGatewayLogGroup = new cwLogs.LogGroup(stack, `${stackName}-apigateway`,
    {
      logGroupName: `${stackName}-api-gateway`,
      retention: cwLogs.RetentionDays.ONE_MONTH,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    }
  );

  const api = new apigateway.RestApi(stack, `${stackName}-api`, {
    domainName: {
      domainName: subdomain,
      certificate,
    },
    deployOptions: {
      methodOptions: {
        '/*/*': {  // This special path applies to all resource paths and all HTTP methods
          throttlingRateLimit: 100,
          throttlingBurstLimit: 1000
        }
      },
      accessLogDestination: new apigateway.LogGroupLogDestination(apiGatewayLogGroup),
      accessLogFormat: apigateway.AccessLogFormat.jsonWithStandardFields(
        {
          caller: false,
          httpMethod: true,
          ip: true,
          protocol: true,
          requestTime: true,
          resourcePath: true,
          responseLength: true,
          status: true,
          user: true
        }
      )
    }
  });
  api.node.addDependency(fargateService);

  api.root.addMethod('ANY', new apigateway.HttpIntegration(targetURL, { httpMethod: 'ANY' }));
  api.root.addProxy({
    defaultMethodOptions: {
      requestParameters: { "method.request.path.proxy": true }
    },
    defaultIntegration: 
    new apigateway.HttpIntegration(`${targetURL}/{proxy}`, {
      httpMethod: 'ANY',
      options: {
        requestParameters: {
          "integration.request.path.proxy": "method.request.path.proxy",
        },
        cacheKeyParameters: [
          "method.request.path.proxy"
        ],
        integrationResponses: [
          {
            statusCode: "200"
          }
        ]
      },
      proxy: true,
    }),
  });

  new route53.ARecord(stack, 'CustomDomainAliasRecord', {
    zone: hostedZone,
    target: route53.RecordTarget.fromAlias(new targets.ApiGateway(api)),
    recordName: subdomain,
  });
}