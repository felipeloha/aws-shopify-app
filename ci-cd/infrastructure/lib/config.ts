export const envSecretsConfig = {
  secretKeys: {
    shopifyApiKey: "SHOPIFY_API_KEY",
    shopifyApiSecret: "SHOPIFY_API_SECRET",
    shop: "SHOP",
    scopes: "SCOPES",
    host: "HOST",
  },
};
