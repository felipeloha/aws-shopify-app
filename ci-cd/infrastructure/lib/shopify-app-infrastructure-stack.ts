import * as cdk from '@aws-cdk/core';
import * as ec2 from '@aws-cdk/aws-ec2';
import * as ecs from '@aws-cdk/aws-ecs';
import * as ecs_patterns from '@aws-cdk/aws-ecs-patterns';
import * as cwLogs from '@aws-cdk/aws-logs';
import * as ecr from '@aws-cdk/aws-ecr';
import * as sm from "@aws-cdk/aws-secretsmanager";
import createApiGateway from './api-gateway';
import { envSecretsConfig } from "./config";

export class ShopifyAppInfrastructureStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const version = new cdk.CfnParameter(this, "version", {
      description: "Branch + commit",
      type: "String",
      default: "develop-df47282",
    });

    const stackName = cdk.Stack.of(this).stackName;

    //ecs
    const vpc = ec2.Vpc.fromLookup(this, "vpc", { isDefault: true });

    const cluster = new ecs.Cluster(this, "ShopifyCluster", {
      clusterName: `${stackName}-cluster`,
      vpc: vpc
    });

    const ecsLogGroup = new cwLogs.LogGroup(this, `${stackName}-ecs`, {
      logGroupName: `${stackName}-ecs`,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
      retention: cwLogs.RetentionDays.ONE_MONTH,
    });

    const ecsLogDriver = new ecs.AwsLogDriver({
      logGroup: ecsLogGroup,
      streamPrefix: `${stackName}-logs`,
    });

    const appSecrets = sm.Secret.fromSecretNameV2(this, 'ShopifyAppSecret', `${stackName}-secrets`);

    // Create a load-balanced Fargate service and make it public
    const fargateService = new ecs_patterns.ApplicationLoadBalancedFargateService(this, "ShopifyAppService", {
      cluster: cluster,
      //TODO is this a security risk? actually we dont want this
      assignPublicIp: true,
      propagateTags: ecs.PropagatedTagSource.SERVICE,
      serviceName: `${stackName}-service`,
      taskImageOptions: {
        containerName: `${stackName}-container`,
        enableLogging: true,
        logDriver: ecsLogDriver,
        image: ecs.ContainerImage.fromEcrRepository(
          ecr.Repository.fromRepositoryName(this, 'shopify-app', 'shopify-app'),
          version.valueAsString,
        ),
        environment: {
          SHOPIFY_APP_VERSION: version.valueAsString,
        },
        secrets: {
          SHOPIFY_API_KEY: ecs.Secret.fromSecretsManager(
            appSecrets,
            envSecretsConfig.secretKeys.shopifyApiKey
          ),
          SHOPIFY_API_SECRET: ecs.Secret.fromSecretsManager(
            appSecrets,
            envSecretsConfig.secretKeys.shopifyApiSecret
          ),
          SHOP: ecs.Secret.fromSecretsManager(
            appSecrets,
            envSecretsConfig.secretKeys.shop
          ),
          SCOPES: ecs.Secret.fromSecretsManager(
            appSecrets,
            envSecretsConfig.secretKeys.scopes
          ),
          HOST: ecs.Secret.fromSecretsManager(
            appSecrets,
            envSecretsConfig.secretKeys.host
          ),
        },
        containerPort: 8081,
      },
      publicLoadBalancer: true,
    });

    fargateService.targetGroup.configureHealthCheck({
      path: "/health",
      interval: cdk.Duration.seconds(10),
    });

    // reduce default from 5min to 30s to speed up deployments
    fargateService.targetGroup.setAttribute("deregistration_delay.timeout_seconds", "30");

    cdk.Tags.of(fargateService).add('Name', `${stackName}-fargate-service`);
    cdk.Tags.of(fargateService.targetGroup).add('Name', `${stackName}-target-group`);
    cdk.Tags.of(fargateService.listener.loadBalancer).add('Name', `${stackName}-load-balancer`);

    createApiGateway(this, stackName, fargateService);
  }
}
