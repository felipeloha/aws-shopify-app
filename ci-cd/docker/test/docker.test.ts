import { expect as expectCDK, haveResource } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import * as Docker from '../lib/docker-stack';
import { SynthUtils } from '@aws-cdk/assert';

test('Empty Stack', () => {
  const app = new cdk.App();
  // WHEN
  const stack = new Docker.DockerStack(app, 'MyTestStack');
  // THEN
  expectCDK(stack).to(haveResource('AWS::CodeBuild::Project',{}));
  expectCDK(stack).to(haveResource('AWS::IAM::Role',{}));
  expectCDK(stack).to(haveResource('AWS::IAM::Policy',{}));
  expectCDK(stack).to(haveResource('AWS::ECR::Repository',{}));
});
test('Snapshot', () => {
const app = new cdk.App();
const stack = new Docker.DockerStack(app, 'TestStack');
expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
