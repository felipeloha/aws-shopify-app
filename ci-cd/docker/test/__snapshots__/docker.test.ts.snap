// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
Object {
  "Resources": Object {
    "CodebuildEE61DA16": Object {
      "Properties": Object {
        "Artifacts": Object {
          "Type": "NO_ARTIFACTS",
        },
        "EncryptionKey": "alias/aws/s3",
        "Environment": Object {
          "ComputeType": "BUILD_GENERAL1_SMALL",
          "Image": "aws/codebuild/amazonlinux2-x86_64-standard:3.0",
          "ImagePullCredentialsType": "CODEBUILD",
          "PrivilegedMode": true,
          "Type": "LINUX_CONTAINER",
        },
        "Name": "shopify-app-docker",
        "ServiceRole": Object {
          "Fn::GetAtt": Array [
            "codebuildRole92FB11B8",
            "Arn",
          ],
        },
        "Source": Object {
          "BuildSpec": "ci-cd/docker/buildspec.yaml",
          "Location": "https://bitbucket.org/mybitbucket/shopify-app.git",
          "ReportBuildStatus": false,
          "Type": "BITBUCKET",
        },
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "codebuild-shopify-app-docker",
          },
        ],
        "TimeoutInMinutes": 10,
      },
      "Type": "AWS::CodeBuild::Project",
    },
    "codebuildRole92FB11B8": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "codebuild.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "codebuildRoleDefaultPolicy47F57BA5": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "ecr:GetAuthorizationToken",
                "ecr:BatchCheckLayerAvailability",
                "ecr:GetDownloadUrlForLayer",
                "ecr:GetRepositoryPolicy",
                "ecr:DescribeRepositories",
                "ecr:ListImages",
                "ecr:DescribeImages",
                "ecr:BatchGetImage",
                "ecr:PutImage",
                "ecr:InitiateLayerUpload",
                "ecr:UploadLayerPart",
                "ecr:CompleteLayerUpload",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":log-group:/aws/codebuild/",
                      Object {
                        "Ref": "CodebuildEE61DA16",
                      },
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":log-group:/aws/codebuild/",
                      Object {
                        "Ref": "CodebuildEE61DA16",
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "codebuild:CreateReportGroup",
                "codebuild:CreateReport",
                "codebuild:UpdateReport",
                "codebuild:BatchPutTestCases",
                "codebuild:BatchPutCodeCoverages",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":codebuild:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":report-group/",
                    Object {
                      "Ref": "CodebuildEE61DA16",
                    },
                    "-*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "codebuildRoleDefaultPolicy47F57BA5",
        "Roles": Array [
          Object {
            "Ref": "codebuildRole92FB11B8",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "shopifyapp52832147": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "LifecyclePolicy": Object {
          "LifecyclePolicyText": "{\\"rules\\":[{\\"rulePriority\\":1,\\"selection\\":{\\"tagStatus\\":\\"untagged\\",\\"countType\\":\\"sinceImagePushed\\",\\"countNumber\\":1,\\"countUnit\\":\\"days\\"},\\"action\\":{\\"type\\":\\"expire\\"}},{\\"rulePriority\\":2,\\"selection\\":{\\"tagStatus\\":\\"tagged\\",\\"tagPrefixList\\":[\\"PUL\\",\\"develop\\",\\"master\\"],\\"countType\\":\\"sinceImagePushed\\",\\"countNumber\\":30,\\"countUnit\\":\\"days\\"},\\"action\\":{\\"type\\":\\"expire\\"}}]}",
        },
        "RepositoryName": "shopify-app",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "shopify-app",
          },
        ],
      },
      "Type": "AWS::ECR::Repository",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
